### PR TITLE
added newton-backtrack

### PR DESCRIPTION
**Pull request by arthbous**
I started the newton backtrack as a function, in newton_func.cpp.

I don't like it because it is too specific to pnp...I have to include the pnp.h obtained from the ufl file. So maybe it should be in the main file, but that's a lot of lines..

### CODE REVIEW

**Comment by maximilianmetti** [comment]
I see what you mean... I have some ideas for when we expand the type of PNP equations that we'll need to handle coefficients (and the number of charged species) in a more generalized context.  Maybe then, we can add this functionality.

I haven't looked at it yet, but it should be okay to have some PNP specific functions.  Maybe we should be careful to include it in the appropriate library.

File: src/newton_functs.cpp [hunk 1]
```diff
@@ -31,6 +31,48 @@ void update_solution (dolfin::Function* iterate, dolfin::Function* update)
...
+    update_interpolant.interpolate(*update);
+    dolfin::Function _iterate(iterate);
+    *(_iterate->vector()) += *(update_interpolant.vector());
+    L.CatCat = _iterate[0];
+    L.AnAn   = _iterate[1];
+    L.EsEs   = _iterate[2];
+    assemble(*b, L);
+    bc.apply(*b);
+    new_residual = (*b).norm("l2");
+    while ( (new_residual>relative_residual) and i<10)
```
Comment (maximilianmetti) [inline src/newton_functs.cpp]: we shouldn't hard code parameters like this 10. It should be named ``maxDamps`` for clarity (or some other name that describes it).

File: src/newton_functs.cpp [hunk 2]
```diff
@@ -31,6 +31,48 @@ void update_solution (dolfin::Function* iterate, dolfin::Function* update)
...
+    {
+      *(_iterate->vector()) = *(iterate.vector());
+      *(_iterate->vector()) += (0.5)**i*(update_interpolant.vector());
```
Comment (maximilianmetti) [inline src/newton_functs.cpp]: The dampFactor should also be named.

**Comment by arthbous** [comment]
I agree. I am still thinking on where to add this function because it need pnp.h, because of L.CatCat and so on.

**Pull request closed without merging**